File: packages/cli/src/commands/get.ts
import { Command } from "commander";
import chalk from "chalk";
import ora from "ora";
import { createApiClient } from "../lib/api-client.js";
import {
  formatEntries,
  info,
  type OutputFormat,
} from "../lib/output.js";
import {
  resolveProject,
  handleError,
  validateEnum,
  validatePositiveInt,
} from "../lib/shared.js";

const VALID_FORMATS = ["json", "table", "markdown"];

export const getCommand = new Command("get")
  .description("Retrieve context entries with optional filtering")
  .option("-p, --project <slug>", "Project slug (reads from .remb.yml if omitted)")
  .option("-f, --feature <name>", "Filter by feature name")
  .option("-l, --limit <n>", "Max entries to return", "10")
  .option("--format <format>", "Output format: json, table, markdown", "table")
  .addHelpText(
    "after",
    `\nExamples:\n  $ remb get -f auth\n  $ remb get -p my-app --format json\n  $ remb get -l 50 --format markdown`,
  )
  .action(async (opts) => {
    const projectSlug = resolveProject(opts.project);
    const limit = parseInt(opts.limit, 10) || 10;
    validatePositiveInt(limit, "Limit", 200);
    validateEnum(opts.format, "format", VALID_FORMATS);
    const format = opts.format as OutputFormat;

    const spinner = ora("Fetching context...").start();

    try {
      const client = createApiClient();
      const result = await client.getContext({
        projectSlug,
        featureName: opts.feature,
        limit,
      });

      spinner.stop();

      if (result.entries.length === 0) {
        info(
          opts.feature
            ? `No entries found for feature ${chalk.bold(opts.feature)} in ${chalk.bold(projectSlug)}.`
            : `No entries found for project ${chalk.bold(projectSlug)}.`
        );
        return;
      }

      console.log(formatEntries(result.entries, format));

      if (format !== "json") {
        console.log();
        info(
          chalk.dim(`Showing ${result.total} entries.`) +
            (result.total >= limit
              ? chalk.dim(` Use --limit to see more.`)
              : "")
        );
      }
    } catch (err) {
      spinner.stop();
      handleError(err);
    }
  });

File: packages/cli-go/cmd/get.go
package cmd

import (
	"fmt"
	"os"
	"strconv"

	"github.com/useremb/remb/internal/api"
	"github.com/useremb/remb/internal/output"
	"github.com/spf13/cobra"
)

var getProject string
var getFeature string
var getLimit string
var getFormat string

var getCmd = &cobra.Command{
	Use:   "get",
	Short: "Retrieve context entries with optional filtering",
	RunE:  runGet,
}

func init() {
	getCmd.Flags().StringVarP(&getProject, "project", "p", "", "Project slug (reads from .remb.yml if omitted)")
	getCmd.Flags().StringVarP(&getFeature, "feature", "f", "", "Filter by feature name")
	getCmd.Flags().StringVarP(&getLimit, "limit", "l", "10", "Max entries to return")
	getCmd.Flags().StringVar(&getFormat, "format", "table", "Output format: json, table, markdown")
}

func runGet(cmd *cobra.Command, args []string) error {
	projectSlug := resolveProject(getProject)
	limit, err := strconv.Atoi(getLimit)
	if err != nil || limit <= 0 {
		limit = 10
	}

	fmt.Print("⠋ Fetching context...")

	client, err := api.NewClient()
	if err != nil {
		fmt.Print("\r\033[K")
		output.Error(err.Error())
		os.Exit(1)
	}

	result, err := client.GetContext(projectSlug, getFeature, limit)

	fmt.Print("\r\033[K")

	if err != nil {
		handleAPIError(err)
	}

	if len(result.Entries) == 0 {
		if getFeature != "" {
			output.Info(fmt.Sprintf("No entries found for feature %s in %s.",
				output.Bold(getFeature), output.Bold(projectSlug)))
		} else {
			output.Info(fmt.Sprintf("No entries found for project %s.", output.Bold(projectSlug)))
		}
		return nil
	}

	// Convert to output entries
	entries := make([]output.Entry, len(result.Entries))
	for i, e := range result.Entries {
		entries[i] = output.Entry{
			ID:        e.ID,
			Feature:   e.Feature,
			Content:   e.Content,
			EntryType: e.EntryType,
			Source:    e.Source,
			CreatedAt: e.CreatedAt,
		}
	}

	fmt.Println(output.FormatEntries(entries, getFormat))

	if getFormat != "json" {
		fmt.Println()
		msg := fmt.Sprintf("Showing %d entries.", result.Total)
		if result.Total >= limit {
			msg += " Use --limit to see more."
		}
		output.Info(output.Dim(msg))
	}

	return nil
}
